alacritty/src/renderer/box_drawing.rs

```diff
@@ -0,0 +1,584 @@
+        for x in from_x as usize..=to_x as usize {
+            let y = from_y + d_y * (x as f32 - from_x) / d_x;
+            let y = y.clamp(0., self.height as f32 - 1.);
+            let index = min(x + y as usize * self.width, self.buffer.len() - 1);
+            self.buffer[index] = COLOR_FILL;
+        }
+    }
+
+    /// Draws an arc from `(0, self.center_v())` to `(self.center_h(), 0)`.
+    ///
+    /// You can mirror Arc in whichever direction you'd like.
+    fn draw_arc_centered(&mut self, stroke_size: usize) {
+        let v_line_bounds = self.v_line_bounds(self.center_h(), stroke_size);
+        let v_line_bounds = (v_line_bounds.0 as usize + 1, v_line_bounds.1 as usize);
+        let h_line_bounds = self.h_line_bounds(self.center_v(), stroke_size);
+        let h_line_bounds = (h_line_bounds.0 as usize + 1, h_line_bounds.1 as usize);
+
+        for (to_x, from_y) in (v_line_bounds.0..=v_line_bounds.1)
+            .into_iter()
+            .zip((h_line_bounds.0..=h_line_bounds.1).into_iter())
+        {
+            let d1 = to_x as f32;
+            let d2 = from_y as f32;
+
+            let mut y = from_y as f32;
+            while y >= 0. {
+                let x = f32::sqrt(d2 * d2 - y * y) * d1 / d2;
+                let x_r = min(x as usize, v_line_bounds.1 as usize - 1);
+                let y_r = min(y as usize, h_line_bounds.1 as usize - 1);
+                let index = min(x_r + y_r as usize * self.width, self.buffer.len() - 1);
+                self.buffer[index] = COLOR_FILL;
+                y -= 0.1;
+            }
+        }
+    }
+
+    /// Fills the `Canvas` with the given `Color`.
+    fn fill(&mut self, color: RGBPixel) {
+        self.buffer.fill(color);
+    }
+
+    /// Consumes `Canvas` and returns its underlying storage as raw byte vector.
+    fn into_raw_buffer(self) -> Vec<u8> {
+        // SAFETY This is safe since we use `repr(packed)` on `Pixel` struct for underlying storage
+        // of the `Canvas` buffer which consists of three u8 values.
+        unsafe {
+            let capacity = self.buffer.capacity();
+            let len = self.buffer.len() * 3;
+            let buf = self.buffer.as_ptr() as *mut u8;
+            std::mem::forget(self.buffer);
+            Vec::from_raw_parts(buf, len, capacity)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crossfont::Metrics;
+
+    #[test]
+    fn builtin_line_drawing_glyphs_coverage() {
+        // Dummy metrics values to test builtin glyphs coverage.
+        let metrics = Metrics {
+            average_advance: 6.,
+            line_height: 16.,
+            descent: 4.,
+            underline_position: 2.,
+            underline_thickness: 2.,
+            strikeout_position: 2.,
+            strikeout_thickness: 2.,
+        };
+
+        // Test coverage of box drawing characters.
+        for character in '\u{2500}'..='\u{259f}' {
+            assert!(builtin_glyph(character, &metrics).is_some());
```

Comment:
Might be worthwhile to check the negative case too maybe? Not sure what range you'd pick for that, but maybe something surrounding the box drawing range just to make sure?

Reply:
The thing is that if something is missing in the function it'll just straight crash on some `unreachable!()`, so I'm mostly testing those things that nothing is actually missed here. I'm not sure that for negative it makes sense.

Reply:
The `builtin_glyph` function should never panic, right?

Reply:
box_drawing could panic if you remove some element from pattern matching.

Reply:
I mean in that case it would be good to have a test that fails, right?